Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/DefaultIfEmpty.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT License.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+namespace Reaqtive.Operators
+{
+    internal sealed class DefaultIfEmpty<TSource> : SubscribableBase<TSource>
+    {
+        private readonly ISubscribable<TSource> _source;
+
+        public DefaultIfEmpty(ISubscribable<TSource> source)
+        {
+            Debug.Assert(source != null);
+
+            _source = source;
+        }
+
+        protected override ISubscription SubscribeCore(IObserver<TSource> observer)
+        {
+            return new _(this, observer);
+        }
+
+        private sealed class _ : StatefulUnaryOperator<DefaultIfEmpty<TSource>, TSource>, IObserver<TSource>
+        {
+            private bool _isNotEmpty;
+
+            public _(DefaultIfEmpty<TSource> parent, IObserver<TSource> observer)
+                : base(parent, observer)
+            {
+            }
+
+            public override string Name => "rc:DefaultIfEmpty";
+
+            public override Version Version => Versioning.v1;
+
+            public void OnCompleted()
+            {
+                if (!_isNotEmpty)
+                {
+                    Output.OnNext(default);
+                }
+                Output.OnCompleted();
+                Dispose();
+            }
+
+            public void OnError(Exception error)
+            {
+                Output.OnError(error);
+                Dispose();
+            }
+
+            public void OnNext(TSource value)
+            {
+                _isNotEmpty |= (StateChanged = true);
```

Comment:
This does not short-circuit unlike `||`, so we'll always assign to `StateChanged` and thus mark the operator state as dirty unnecessarily (which in turn will result in extra checkpointing activity). We should change it to:

```csharp
if (!_isNotEmpty)
{
    _isNotEmpty = true;
    StateChanged = true;
}
```

Note that operators start off as "dirty" when they get created, so the first time we checkpoint, we'll save the `false` value (unless, of course, there's already been an event in between creation and checkpointing). When receiving an event, we'll flip the flag to `true` once, and incur at most one future checkpoint of this operator's state.